Services/Infrastructure/OJS.Services.Infrastructure/Cache/Implementations/CacheService.cs

```diff
@@ -1,236 +1,207 @@
+                return Task.CompletedTask;
+            },
+            () => Task.FromResult(Task.CompletedTask),
+            $"Remove_{cacheId}");
 
-            try
-            {
-                this.cache.Remove(cacheId);
-            }
-            catch (RedisCommandException ex)
-            {
-                this.emailService.SendEmail(this.emailConfig.DevEmail, GetExceptionTypeAsString(ex), ex.Message);
-            }
+    private static async Task<T> ExecuteWithResilienceContext<T>(Func<ResilienceContext, Task<T>> action)
+    {
+        var context = ResilienceContextPool.Shared.Get();
+        try
+        {
+            return await action(context);
         }
-
-        private static T ParseValue<T>(byte[] valueAsByteArray)
-            => new StreamReader(new MemoryStream(valueAsByteArray))
-                .ReadToEnd()
-                .FromJson<T>();
-
-        private static byte[] ParseValue<T>(T obj)
-            => obj!
-                .ToJson()
-                .ToByteArray();
-
-        private static string GetExceptionTypeAsString(Exception exception) => exception.GetType().ToString();
-
-        private DateTime GetAbsoluteExpirationByCacheSeconds(int cacheSeconds)
-            => this.dates
-                .GetUtcNow()
-                .AddSeconds(cacheSeconds);
-
-        private async Task VerifyValueInCache<T>(
-            string cacheId,
-            Func<Task<T>> getItemCallback,
-            DateTime absoluteExpiration)
+        finally
         {
-            var value = await this.cache.GetAsync(cacheId);
-            if (value.IsNull())
-            {
-                var options = new DistributedCacheEntryOptions()
-                    .SetAbsoluteExpiration(absoluteExpiration);
-
-                var result = await getItemCallback();
-                await this.cache.SetAsync(
-                    cacheId,
-                    ParseValue(result),
-                    options);
-            }
+            ResilienceContextPool.Shared.Return(context);
         }
+    }
 
-        private async Task<T> GetItemResultAsync<T>(Func<Task<T>> resultAction, Func<Task<T>> fallbackResultAction)
+    private static T ParseValue<T>(byte[] valueAsByteArray)
+        => new StreamReader(new MemoryStream(valueAsByteArray))
+            .ReadToEnd()
+            .FromJson<T>();
+
+    private static byte[] ParseValue<T>(T obj)
+        => obj!
+            .ToJson()
+            .ToByteArray();
+
+    private async Task<T> ExecuteWithCircuitBreaker<T>(
+        Func<ResilienceContext, Task<T>> action,
+        Func<Task<T>> fallbackAction,
+        string operationKey)
+        => await ExecuteWithResilienceContext(async (context) =>
         {
-            if (!this.IsRedisConnected())
-            {
-                if (this.ShouldSendExceptionEmail(EmailConstants.RedisSubject, EmailConstants.RedisBody))
+            context.Properties.Set(new ResiliencePropertyKey<string>("OperationKey"), operationKey);
```

Comment:
This logic too can go into the common service along with the `operationKey` (which should be added to a constant, as it's used in the listener also)

Reply:
Done.